src/Embedders/Word2Vec.php

```
@@ -0,0 +1,937 @@
+        }
+
+        if ($epochs < 1) {
+            throw new InvalidArgumentException("Number of epochs must be greater than 0, $epochs given.");
+        }
+
+        if ($minCount < 1) {
+            throw new InvalidArgumentException("Minimum word count must be greater than 0, $minCount given.");
+        }
+
+        $this->layer = $layer;
+        $this->window = $window;
+        $this->dimensions = $dimensions;
+        $this->sampleRate = $sampleRate;
+        $this->alpha = $alpha;
+        $this->epochs = $epochs;
+        $this->minCount = $minCount;
+        $this->neg_labels = Vector::quick([1, 0]);
+    }
+
+    /**
+     * Return the data types that this estimator is compatible with.
+     *
+     * @return \Rubix\ML\DataType[]
+     */
+    public function compatibility() : array
+    {
+        return [
+            DataType::categorical(),
+        ];
+    }
+    
+    /**
+     * Return the settings of the hyper-parameters in an associative array.
+     *
+     * @return mixed[]
+     */
+    public function params() : array
+    {
+        return [
+            'layer' => $this->layer,
+            'window' => $this->window,
+            'dimensions' => $this->dimensions,
+            'sample_rate' => $this->sampleRate,
+            'alpha' => $this->alpha,
+            'epochs' => $this->epochs,
+            'min_count' => $this->minCount,
+        ];
+    }
+
+    /**
+     * Has the learner been trained?
+     *
+     * @return bool
+     */
+    public function trained() : bool
+    {
+        return !empty($this->vectors);
+    }
+
+    /**
+     * Iterating through a range of the specific epoch count and updating all respective word vectors.
+     * @param \Rubix\ML\Datasets\Dataset $dataset
+     * @throws \InvalidArgumentException
+     */
+    public function train(Dataset $dataset) : void
+    {
+        DatasetIsNotEmpty::check($dataset);
+
+        $sentences = $dataset->column(0);
```

What happens if the dataset contains more columns? What if the data in the first column is invalid?

There should be more validation here and if the intent is to only support one vocabulary, it should be known to the user by way of an exception message

---

This kind of deals with what we discussed the other day: a user trying to embed on a multi-feature dataset. 

The only check I can think of is if the contents of the feature are of type string.

---

We're going to need to come up with a way to handle this in order for it to be compatible with the Embedder/Transformer interface

---

So we still need to implement a way to use (or at least handle) *all* columns of the dataset. For example, the user may have multiple text columns.

---

But what if one of the features are not text or they do not want it embedded?

---

There shouldn't be multiple features per sample in a word2vec dataset. This is set up to process one feature of sentences. I can add that in the documentation. But if we want to change it then we need to have a serious conversation. That would require a lot of work and isn't how word2vec is used.

---

Since we decided to only support one text feature, we need to throw an exception if the user tries to fit a dataset with more than 1 column in the samples.

---

Already handled.